### PR TITLE
2 frames 2 pacing

### LINKED ISSUE #1072: Window stuttering after these commits
Commits 86d37393...a4fae2b6 created a stuttering when moving windows, which is very noticeable on the 144hz monitor.

The problem only disappeared when I removed these three commits. Just deleted commits one by one and watched as the lag disappeared.

Here's a video recorded at 120fps that shows this issue.

[2023-06-13_16-58-50.mkv.zip](https://github.com/yshui/picom/files/11736351/2023-06-13_16-58-50.mkv.zip)

### PR DESCRIPTION
**Pull request by yshui**
Another batch of frame pacing improvements. Hopefully it's acceptable for general use now.

Fixes #1072

### CODE REVIEW

**Comment by r0-zero** [comment]
@yshui 

Hi,

I'm not much aware of the frame pacing topic but when enabled with animations (glx, xrender) everything seems to be messed up. After opening several windows any window movement/resize becomes too laggy and crashes everything eventually (all windows disappear or start making awkward movements). 

Can you guide me a little bit about frame pacing? I'm not sure if it should be included with animations though. Also my fork is not like other forks, it is 100% original picom (this one) with animations included in it. I also maintain all of your commits, so that it is not behind of the original.


```
	if ((ps->o.legacy_backends || ps->o.animations || ps->o.benchmark || !ps->backend_data->ops->last_render_time) &&
	    ps->frame_pacing) {
		// Disable frame pacing if we are using a legacy backend or if we are in
		// benchmark mode, or if the backend doesn't report render time
```

Also if you have a little bit of free time I have one more question, it is not related to this issue.
Since animations (resize/window movements) trigger hundreds of events (clear window flags/set window flags/add damage etc) in a second I think this increases CPU usage and decreases performance a bit; since if animation is not used it will trigger only once per one event. Or for example, picom spams logs because it tries to clear flags of a destroyed window (it fades out and triggers a custom animation) but window is unmapped/destroyed from X. Previously I just blocked sending logs if the window animation is still in process but it is not an ideal solution, since it only blocks sending log messages, I wonder if there is any way to do it without those functions being called?

**Comment by yshui** [comment]
@FT-Labs can you try merging this branch with your animation branch?

**Comment by yshui** [comment]
@FT-Labs btw, what's the current status on animation? judging from the PRs there are multiple efforts on animation, and i don't know how they are related to each other.

**Comment by r0-zero** [comment]
@yshui 

I've merged the code with this branch and it fixed all performance issues. There is no performance issue with frame pacing enabled now.

https://github.com/FT-Labs/picom/tree/pacing-fixes

It is close to done actually, I do not have that much idea about other forks. As a start, I have used @dccsillag's fork 1 1-5 years ago, but a lot has changed since it did not get updated. I just try to keep in line with original picom without breaking the animations.

To sum up, here is what works now:
Shadows without glitches
Smooth animations without using too much cpu/memory
Fade out destroyed/unmapped windows
~10 different animations
Open/close window animations
Workspace switch animations
Works with custom shaders 
No flickering on any transparent window (focused/not focused) when resizing/moving

Remaining problems:
Make workspace switch animation behave correctly for any WM/DE 
 - This is troublesome because workspace/tag/desktops are dependent on developers, and there is no strict rule. However, the preferred way is set all of NET_WM_DESKTOP of all windows then change NET_CURRENT_DESKTOP atom of root window. The issue is while awesomewm or xmonad follows this rule but other wm like dwm have different implementations
Sometimes workspace switch animation can go towards center of root viewport instead of going to the center of current monitor. This issue only occurs on multiple monitor setups. Like minimize animation shrinks all of the visible windows to the center of monitor. 
Some people reported glitches with some layouts in bspwm, most of them are solved with disabling pacing before this branch. 

Aside from these it is done and works good. Of course, there could be more improvements but I think they can be done later since they would classify as features

**Comment by yshui** [comment]
@FT-Labs cool. after 11 is out of the door, the focus for 12 would be getting animation over the finishing line

**Comment by Stebalien** [comment]
I'm no longer noticing any power draw differences introduced by frame pacing when running this branch, but I've also switched from Intel to AMD so I don't get quite the same level of power state information 🤷.

**Comment by r0-zero** [comment]
@yshui sorry if it is a stupid question, what is 11 and 12 :D

**Comment by absolutelynothelix** [comment]
@FT-Labs, picom v11 and v12 respectively.

**Comment by r0-zero** [comment]
Ok, got it

**Comment by morgoth6** [comment]
It seems if _smart_frame_pacing_ is enabled, picom _log-spam_ huge amount of warnings:
```
PresentCompleteNotify msc is going backwards, last_msc: 767522245 , current msc: 0
```
As soon as for display turns off (DPMS) In my case my  `picom.log` grow to 2GB+ (and eats quite significant portion of my battera at the same time) 

Maybe it would be good idea to utilize `DPMSInfoNotify` ? It should be working now ([#134](https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/134)) and for older systems it would be no-op (?)

File: src/picom.c [hunk 1]
```diff
@@ -142,43 +145,158 @@ static inline struct managed_win *find_win_all(session_t *ps, const xcb_window_t
...
+				          "%" PRIu64 ", not adding sample.",
+				          frame_count, frame_time, e->ust);
+			}
+		}
+		ps->last_msc_instant = e->ust;
+		ps->last_msc = e->msc;
+	} else if (ps->last_msc > e->msc) {
+		log_warn("PresentCompleteNotify msc is going backwards, last_msc: "
+		         "%" PRIu64 ", current msc: %" PRIu64,
+		         ps->last_msc, e->msc);
```
Comment (morgoth6) [inline src/picom.c:224]: I think if we hit this spot once we are in trouble as `ps->last_msc` stays the same and next collect_vblank_interval_statistics() would land in the same place unconditionally (also see my comment [1868267108](https://github.com/yshui/picom/pull/1156#issuecomment-1868267108))

I adjusted this part for testing (incorrectly propably, but just to test _smart_frame_pacing_) to something like:
```
@@ -222,6 +222,15 @@ collect_vblank_interval_statistics(struct vblank_event *e, void *ud) {
        log_warn("PresentCompleteNotify msc is going backwards, last_msc: "
                 "%" PRIu64 ", current msc: %" PRIu64,
                 ps->last_msc, e->msc);
       ps->last_msc_instant = e->ust;
       ps->last_msc = e->msc;
       return VBLANK_CALLBACK_AGAIN;
   }
   else if(e->msc == 0) {
       // DPMS ?
       ps->last_msc_instant = 0;
       ps->last_msc = 0;
       return VBLANK_CALLBACK_DONE;
    }
 
    vblank_interval = render_statistics_get_vblank_time(&ps->render_stats);

```

**Comment by yshui** [comment]
Wow, `DPMSInfoNotify` is merged?! Never thought I'd see this day :laughing:

File: src/picom.c (hunk 1 above)
Comment (yshui) [reply src/picom.c:224]: Yes, I would do something similar to this as well.

**Comment by yshui** [comment]
Things seem to be going well, so I am merging this.

**Pull request merged**